### ISSUE TITLE
bind-dirs wrong user / group permissions

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#8466]
### Qubes OS release

R4.2

### Brief summary

User / group permissions in [Qubes `bind-dirs` directory](https://www.qubes-os.org/doc/bind-dirs/) are wrong.

### Steps to reproduce

Not sure. Might happen by swapping the the underlying Template of a App Qube for another Template or by restoring a VM using Qubes VM backup (which could be similar to the former).

### Expected behavior

Correct user / group permissions.

### Actual behavior

In corner cases, wrong user / group permissions.

### Additional information 

* For tracking purposes: This is not a Whonix specific issue. This is a general issues with Qubes bind-dirs.
* Affects Qubes-Whonix: Yes. In corner cases users have to run a [Tor permissions fix](https://www.whonix.org/wiki/Tor#Permissions_Fix) because Tor fails to start due to wrong folder permissions.

I was now able to most likely pin this issue to Qubes bind-dirs. The permissions in folder `/rw/bind-dirs` (where the fully-persistent data in App Qube resides) (for example "`/rw/bind-dirs/var/lib/tor`") are wrong to begin with. Therefore bind-dirs will mount it also with the wrong permissions as a follow-up error.

Why are the user / group permissions in `/rw/bind-dirs` wrong to begin with? Because while Template A and Template B might use the same user / group names (example: "`debian-tor`"), the user identifier (`UID`) and group identifier (`GID`), which are numeric, are most likely different. Unfortunately, Linux stores UID / GID as number in the filesystem, not as literal names.

**Comment by DemiMarie** [comment]
I don’t think this is fixable in Qubes OS, at least with the current bind-dirs code.  What _could_ be done is to use `systemd-tmpfiles` to fix the permissions on e.g. `/var/lib/tor` as needed.  Correct ordering is guaranteed because `systemd-tmpfiles-setup.service` runs _after_ `local-fs.target`, whereas `qubes-mount-dirs.service` runs _before_ `local-fs.target`.

**Comment by marmarek** [comment]
Can bind-dir apply permissions of the dir it mounts over? For example if `/var/lib/tor` is owned by UID 432, then adjust owner of the mounted dir to the same. The tricky part is what to do about dir content - recursive chown in this case sounds risky (even if incidentally correct in case of tor)...
Maybe there should be opt-in feature for that, for example instead of setting `binds+=( "/var/lib/tor" )`, set `binds_with_chown+=( "/var/lib/tor" )` ? Sounds a bit clumsy...

But in practice, I think solution proposed by @DemiMarie is better. Or alternatively, have `ExecStartPre=chown ...` added to the tor service.

**Comment by adrelanos** [comment]
@DemiMarie 
> What _could_ be done is to use `systemd-tmpfiles` to fix the permissions on e.g. `/var/lib/tor` as needed.

Would probably be best if upstream would ship a systemd-tmpfiles config snippet. Wrote a feature request:
[ship systemd-tmpfiles /usr/lib/tmpfiles.d/tor.conf to fix permission issues](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1050862)

> Correct ordering is guaranteed because `systemd-tmpfiles-setup.service` runs _after_ `local-fs.target`, whereas `qubes-mount-dirs.service` runs _before_ `local-fs.target`.

Ah nice, so that's already implemented.

@marmarek 
> Can bind-dir apply permissions of the dir it mounts over? For example if `/var/lib/tor` is owned by UID 432, then adjust owner of the mounted dir to the same.

That's a cool idea.

> The tricky part is what to do about dir content - recursive chown in this case sounds risky (even if incidentally correct in case of tor)...

I am not aware of such complex cases but in theory it could happen.

> Maybe there should be opt-in feature for that, for example instead of setting `binds+=( "/var/lib/tor" )`, set `binds_with_chown+=( "/var/lib/tor" )` ? Sounds a bit clumsy...

Yes.

> But in practice, I think solution proposed by @DemiMarie is better. Or alternatively, have `ExecStartPre=chown ...` added to the tor service.

Yes.

**Comment by pgerber** [comment]
I also recently ran into this issue when upgrading from Debian bookworm to trixie.

Some time ago, there was an LWN article talking about this issue:

> The Fedora Project is looking for solutions to an interesting problem with its image-based editions and spins, such as the [Atomic Desktops](https://fedoraproject.org/atomic-desktops/) or [CoreOS](https://fedoraproject.org/coreos/), that are created with [rpm-ostree](https://coreos.github.io/rpm-ostree/) or [bootc](https://bootc-dev.github.io/bootc/). If a package that is part of a image-based version has a user or group created dynamically on installation, and it owns files installed on the system, the system may be subject to user ID (UID) and group ID (GID) "drift" on updates. This "UID/GID drift" may come about when a new image with updates is generated, and therefore files may have the wrong ownership. This can have side-effects ranging from mildly inconvenient to serious. No solutions have been adopted just yet, but there are a few ideas on how to deal with the problem.

source: https://lwn.net/Articles/1018082/

It may be worth watching what other affected projects do.

**Comment by adrelanos** [comment]
PR by @ArrayBolt3

* https://github.com/QubesOS/qubes-core-agent-linux/pull/615
